java-checks/src/main/java/org/sonar/java/checks/security/UserEnumerationCheck.java

```diff
@@ -0,0 +1,129 @@
+  private static final String BOOLEAN = "boolean";
+  private static final String STRING = "java.lang.String";
+  private static final String THROWABLE = "java.lang.Throwable";
+
+  private static final MethodMatchers setHideUserMatcher = MethodMatchers.create()
+    .ofSubTypes(ABSTRACT_USER_DETAILS_AUTHENTICATION_PROVIDER)
+    .names(HIDE_USER_NOT_FOUND_EXCEPTIONS)
+    .addParametersMatcher(BOOLEAN)
+    .build();
+
+  private static final MethodMatchers loadUserMatcher = MethodMatchers.create()
+    .ofSubTypes(USER_DETAILS_SERVICE)
+    .names(LOAD_USER_BY_USERNAME)
+    .addParametersMatcher(STRING)
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.METHOD_INVOCATION, Tree.Kind.THROW_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.THROW_STATEMENT)) {
+      checkThrowUsernameNotFoundException(((ThrowStatementTree) tree));
+      return;
+    }
+
+    MethodInvocationTree methodInvocationTree = (MethodInvocationTree) tree;
+    Arguments arguments = methodInvocationTree.arguments();
+    if (arguments.isEmpty()){
+      return;
+    }
+    ExpressionTree firstArgument = arguments.get(0);
+
+    checkHiddenUserNotFoundException(methodInvocationTree, firstArgument);
+    checkLoadUserArgUsedInExceptions(methodInvocationTree, firstArgument);
+  }
+
+  private void checkLoadUserArgUsedInExceptions(MethodInvocationTree methodInvocationTree, ExpressionTree expression) {
+    if (loadUserMatcher.matches(methodInvocationTree) && expression.is(Tree.Kind.IDENTIFIER)) {
+      IdentifierTree identifierTree = (IdentifierTree) expression;
+      identifierTree.symbol().usages()
+        .stream().filter(UserEnumerationCheck::checkParentIsThrowable)
+        .forEach(value -> reportIssue(value, MESSAGE));
+    }
+  }
+
+  private void checkHiddenUserNotFoundException(MethodInvocationTree methodInvocationTree, ExpressionTree expression) {
+    if (setHideUserMatcher.matches(methodInvocationTree) && !expression.asConstant(Boolean.class).orElse(true)) {
+      reportIssue(methodInvocationTree, MESSAGE);
+    }
+  }
+
+  private void checkThrowUsernameNotFoundException(ThrowStatementTree tree) {
+    if (tree.expression().symbolType().is(USERNAME_NOT_FOUND_EXCEPTION) && !isInsideLoadUserByUserName(tree)) {
+      reportIssue(tree.expression(), MESSAGE);
+    }
+  }
+
+  private static boolean checkParentIsThrowable(Tree tree) {
+    Tree current = tree.parent();
+    while (current instanceof ExpressionTree || current instanceof Arguments) {
+      if (current.is(Tree.Kind.NEW_CLASS) && ((NewClassTree) current).symbolType().isSubtypeOf(THROWABLE)) {
+        return true;
+      }
+      current = current.parent();
+    }
+    return false;
+  }
+
+  private static boolean isInsideLoadUserByUserName(Tree tree) {
+    Tree current = tree.parent();
+    while (current instanceof BlockTree || current instanceof MethodTree) {
```

Comment:
I feel this condition is not enough, or will not report only on very trivial case. If I write something like the following, there is going to be 2 FPs:
```
 public static class MyUserDetailsService4 implements UserDetailsService {
    @Override
    public MyUserDetails loadUserByUsername(String s) throws UsernameNotFoundException {
      if (s.equals("yolo"))
        throw new UsernameNotFoundException(""); // Compliant

      switch(s) {
        case "admin":
          throw new UsernameNotFoundException(""); // Compliant
        case "batman":
          break;
      }
      return null;
    }
  }
```
In order to handle this case, I would use a stack of methodTree I'm entering, and just checking the top of the stack is NOT the `loadUserByUsername` method.

For the other _parent loop_, I can see it will anyway stop fast. For this one, it will stop too early, and making it more general will trigger the loop until the parent is null which we want to avoid.

Reply:
I have a small question here.

Firstly, I refactored it to additional method visitor. So it checked each method which wasn't `loadUserByUsername` and reported an issue each time an exception `UserNotFoundException` was found.
Then I realized that this way we're going to check all methods in class. And a chance to meet this exception is quite low. I suppose It's better to check exact throw statement if it is in scope of the `loadUserByUsername` method. The only solution that came to my mind will be 
`private static boolean isInsideLoadUserByUserName(Tree tree) {
    Tree current = tree.parent();
    while (current != null) {
      if (current.is(Tree.Kind.METHOD) && LOAD_USER_MATCHER.matches(((MethodTree) current))) {
        return true;
      }
      current = current.parent();
    }
    return false;
  }`

Is it ok or am I missing something?